File: KantechGatewayApp/Jobs/BaseProcessor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace KantechGatewayApp.KantechGatewayApp.Jobs
{
    public abstract class BaseProcessor : IJobProcessor
    {

        public string Key { get; private set; }
        public string ScheduleName { get; private set; }
        public DateTime LastRunUtc { get; private set; }
        public string LastResult { get; private set; }
        public string Status { get; private set; }

        private string _source;
        private string _target;
        protected string _logSub;
        private string _schedule;
        private string[] _requiredFields;
        private string _jobKey;

        public virtual void Configure(string jobKey)
        {
            string inbound = Infrastructure.JobManager.Config("Root.Inbound");
            string completed = Infrastructure.JobManager.Config("Root.Completed");
            string logs = Infrastructure.JobManager.Config("Root.Logs");
            _source = Infrastructure.FileUtil.Combine(inbound, Infrastructure.JobManager.Config($"Job.{jobKey}.Source"));
            _target = Infrastructure.FileUtil.Combine(completed, Infrastructure.JobManager.Config($"Job.{jobKey}.Target"));
            _logSub = Infrastructure.JobManager.Config($"Job.{jobKey}.LogSub", jobKey);
            _schedule = Infrastructure.JobManager.Config($"Job.{jobKey}.Schedule", "TwiceDaily");
            _requiredFields = Infrastructure.JobManager.Config($"Job.{jobKey}.RequiredFields", "").Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToArray();

            Infrastructure.FileUtil.EnsureDir(_source);
            Infrastructure.FileUtil.EnsureDir(_target);
            Infrastructure.FileUtil.EnsureDir(Path.Combine(logs, _logSub));
            ScheduleName = _schedule;
            _jobKey = jobKey;
        }

        protected virtual bool IsTimeToRun()
        {
            var now = DateTime.Now;
            if (_schedule.Equals("TwiceDaily", StringComparison.OrdinalIgnoreCase))
            {
                string[] times = Infrastructure.JobManager.Config("Schedule.TwiceDailyTimes", "09:00,21:00").Split(',');
                return times.Any(t => SameMinute(now, t));
            }
            else if (_schedule.Equals("Daily", StringComparison.OrdinalIgnoreCase))
            {
                string t = Infrastructure.JobManager.Config("Schedule.DailyTime", "02:00");
                return SameMinute(now, t);
            }
            return false;
        }

        private static bool SameMinute(DateTime now, string hhmm)
        {
            string[] parts = hhmm.Split(':');
            int h = int.Parse(parts[0]);
            int m = int.Parse(parts[1]);
            return now.Hour == h && now.Minute == m;
        }

        public void MaybeRunBySchedule()
        {
            if (IsTimeToRun())
                RunOnce();
        }

        public void RunOnce()
        {
            try
            {
                LastRunUtc = DateTime.UtcNow;
                string pattern = Infrastructure.JobManager.Config("File.Pattern", "*.csv");
                string[] files = Directory.GetFiles(_source, pattern);
                int processedCount = 0;
                string status = "";
                if (files.Length == 0)
                {
                    LastResult = "No files";
                    return;
                }

                foreach (var f in files)
                {
                    try
                    {
                        ProcessOneFile(f);
                        processedCount += 1;
                        // Update progress status after each file
                        status = $"{processedCount} out of {files.Length} processed";
                    }
                    catch (Exception ex)
                    {
                        Infrastructure.Logger.Error($"[{Key}] failed to process file {f}", ex, _logSub);
                    }
                }
                LastResult = $"Processed {files.Length} file(s)";
                Status = status;
            }
            catch (Exception ex)
            {
                Infrastructure.Logger.Error($"[{Key}] run failed", ex, _logSub);
                LastResult = $"Error: {ex.Message}";
            }
        }

        private void ProcessOneFile(string filePath)
        {
            char delim = Convert.ToChar(Infrastructure.JobManager.Config("Csv.Delimiter", ","));
            bool hasHeader = bool.Parse(Infrastructure.JobManager.Config("Csv.HasHeader", "true"));
            var rows = Infrastructure.CsvUtil.ReadRows(filePath, delim, hasHeader);

            foreach (var rf in _requiredFields)
            {
                if (rows.Count > 0 && !rows[0].ContainsKey(rf))
                {
                    throw new ApplicationException($"Missing required field '{rf}' in {Path.GetFileName(filePath)}");
                }
            }

            int ok = 0;
            int err = 0;
            foreach (var r in rows)
            {
                try
                {
                    HandleRow(r);
                    ok += 1;
                }
                catch (Exception ex)
                {
                    err += 1;
                    Infrastructure.Logger.Error($"[{Key}] row failed: {string.Join(", ", r.Select(kv => kv.Key + "=" + kv.Value))}", ex, _logSub);
                }
            }

            if (err == 0)
            {
                Infrastructure.FileUtil.SafeMove(filePath, _target, _jobKey);
            }
            else
            {
                string errDir = Infrastructure.JobManager.Config("Root.Error");
                Infrastructure.FileUtil.SafeMove(filePath, Infrastructure.FileUtil.Combine(errDir, Key), _jobKey);
            }

            Infrastructure.Logger.Info($"[{Key}] {Path.GetFileName(filePath)} -> OK:{ok} ERR:{err}", _logSub);
        }

        protected abstract void HandleRow(Dictionary<string, string> row);
    }
}

File: KantechGatewayApp/Jobs/BaseProcessor.vb
Imports System.IO
Imports KantechGatewayApp.Infrastructure

Namespace KantechGatewayApp.Jobs
    Public MustInherit Class BaseProcessor
        Implements IJobProcessor

        Public ReadOnly Property Key As String Implements IJobProcessor.Key
        Public ReadOnly Property ScheduleName As String Implements IJobProcessor.ScheduleName
        Public ReadOnly Property LastRunUtc As Date Implements IJobProcessor.LastRunUtc
        Public ReadOnly Property LastResult As String Implements IJobProcessor.LastResult
        Public ReadOnly Property Status As String Implements IJobProcessor.Status

        Private _source As String
        Private _target As String
        Protected _logSub As String
        Private _schedule As String
        Private _requiredFields As String()
        Private _jobKey As String

        Public Overridable Sub Configure(jobKey As String) Implements IJobProcessor.Configure
            Dim inbound = Infrastructure.JobManager.Config("Root.Inbound")
            Dim completed = Infrastructure.JobManager.Config("Root.Completed")
            Dim logs = Infrastructure.JobManager.Config("Root.Logs")
            _source = KantechGatewayApp.Infrastructure.FileUtil.Combine(inbound, Infrastructure.JobManager.Config($"Job.{jobKey}.Source"))
            _target = KantechGatewayApp.Infrastructure.FileUtil.Combine(completed, Infrastructure.JobManager.Config($"Job.{jobKey}.Target"))
            _logSub = Infrastructure.JobManager.Config($"Job.{jobKey}.LogSub", jobKey)
            _schedule = Infrastructure.JobManager.Config($"Job.{jobKey}.Schedule", "TwiceDaily")
            _requiredFields = Infrastructure.JobManager.Config($"Job.{jobKey}.RequiredFields", "") _
                              .Split({","c}, StringSplitOptions.RemoveEmptyEntries) _
                              .Select(Function(s) s.Trim()).ToArray()
            KantechGatewayApp.Infrastructure.FileUtil.EnsureDir(_source) : KantechGatewayApp.Infrastructure.FileUtil.EnsureDir(_target) : KantechGatewayApp.Infrastructure.FileUtil.EnsureDir(Path.Combine(logs, _logSub))
            _ScheduleName = _schedule
            _jobKey = jobKey
        End Sub

        Protected Overridable Function IsTimeToRun() As Boolean
            Dim now = DateTime.Now
            If _schedule.Equals("TwiceDaily", StringComparison.OrdinalIgnoreCase) Then
                Dim times = Infrastructure.JobManager.Config("Schedule.TwiceDailyTimes", "09:00,21:00").Split(","c)
                Return times.Any(Function(t) SameMinute(now, t))
            ElseIf _schedule.Equals("Daily", StringComparison.OrdinalIgnoreCase) Then
                Dim t = Infrastructure.JobManager.Config("Schedule.DailyTime", "02:00")
                Return SameMinute(now, t)
            End If
            Return False
        End Function

        Private Shared Function SameMinute(now As DateTime, hhmm As String) As Boolean
            Dim parts = hhmm.Split(":"c)
            Dim h = Integer.Parse(parts(0)) : Dim m = Integer.Parse(parts(1))
            Return now.Hour = h AndAlso now.Minute = m
        End Function

        Public Sub MaybeRunBySchedule() Implements IJobProcessor.MaybeRunBySchedule
            If IsTimeToRun() Then RunOnce()
        End Sub

        Public Sub RunOnce() Implements IJobProcessor.RunOnce
            Try
                _LastRunUtc = Date.UtcNow
                Dim pattern = Infrastructure.JobManager.Config("File.Pattern", "*.csv")
                Dim files = Directory.GetFiles(_source, pattern)
                Dim processedCount As Integer = 0
                Dim status As String = ""
                If files.Length = 0 Then
                    _LastResult = "No files"
                    Return
                End If

                For Each f In files
                    Try
                        ProcessOneFile(f)
                        processedCount += 1
                        ' Update progress status after each file
                        status = $"{processedCount} out of {files.Length} processed"
                    Catch ex As Exception
                        Infrastructure.Logger.Error($"[{Key}] failed to process file {f}", ex, _logSub)
                    End Try
                Next
                _LastResult = $"Processed {files.Length} file(s)"
                _Status = status
            Catch ex As Exception
                Infrastructure.Logger.Error($"[{Key}] run failed", ex, _logSub)
                _LastResult = $"Error: {ex.Message}"
            End Try
        End Sub

        Private Sub ProcessOneFile(filePath As String)
            Dim delim = Convert.ToChar(Infrastructure.JobManager.Config("Csv.Delimiter", ","))
            Dim hasHeader = Boolean.Parse(Infrastructure.JobManager.Config("Csv.HasHeader", "true"))
            Dim rows = Infrastructure.CsvUtil.ReadRows(filePath, delim, hasHeader)

            For Each rf In _requiredFields
                If rows.Count > 0 AndAlso Not rows(0).ContainsKey(rf) Then
                    Throw New ApplicationException($"Missing required field '{rf}' in {Path.GetFileName(filePath)}")
                End If
            Next

            Dim ok As Integer = 0, err As Integer = 0
            For Each r In rows
                Try
                    HandleRow(r)
                    ok += 1
                Catch ex As Exception
                    err += 1
                    Infrastructure.Logger.Error($"[{Key}] row failed: {String.Join(", ", r.Select(Function(kv) kv.Key & "=" & kv.Value))}", ex, _logSub)
                End Try
            Next

            If err = 0 Then
                KantechGatewayApp.Infrastructure.FileUtil.SafeMove(filePath, _target, _jobKey)
            Else
                Dim errDir = Infrastructure.JobManager.Config("Root.Error")
                KantechGatewayApp.Infrastructure.FileUtil.SafeMove(filePath, KantechGatewayApp.Infrastructure.FileUtil.Combine(errDir, Key), _jobKey)
            End If

            Infrastructure.Logger.Info($"[{Key}] {Path.GetFileName(filePath)} -> OK:{ok} ERR:{err}", _logSub)
        End Sub

        Protected MustOverride Sub HandleRow(row As Dictionary(Of String, String))
    End Class
End Namespace